tests/helpers.py

```diff
@@ -160,3 +166,174 @@ def __init__(self, **kwargs):
+        elif isinstance(data_or_exc, type) and isinstance(expected_bytes_or_exc, bytes):
+            deserialization_fail.append((expected_bytes_or_exc, data_or_exc))
+        else:
+            data = cast(Tuple[Any, ...], data_or_exc)
+            exception = cast(type[Exception], expected_bytes_or_exc)
+            kwargs = dict(zip([f[0] for f in fields], data))
+            validation_fail.append((kwargs, exception))
+
+    def generate_name(param: dict[str, Any] | bytes, i: int) -> str:
+        """Generate a name for the test case."""
+        length = 30
+        result = f"{i:02d}] : "  # the first [ is added by pytest
+        if isinstance(param, bytes):
+            try:
+                result += str(param[:length], "utf-8") + "..." if len(param) > (length + 3) else param.decode("utf-8")
+            except UnicodeDecodeError:
+                result += repr(param[:length]) + "..." if len(param) > (length + 3) else repr(param)
+        else:
+            param = cast(Dict[str, Any], param)
+            begin = ", ".join(f"{k}={v}" for k, v in param.items())
+            result += begin[:length] + "..." if len(begin) > (length + 3) else begin
+        result = result.replace("\n", "\\n").replace("\r", "\\r")
+        result += f" [{cls.__name__}"  # the other [ is added by pytest
+        return result
+
+    class TestClass:
+        """Test class for the generated tests."""
+
+        @pytest.mark.parametrize(
+            ("kwargs", "expected_bytes"),
+            parameters,
+            ids=tuple(generate_name(kwargs, i) for i, (kwargs, _) in enumerate(parameters)),
+        )
+        def test_serialization(self, kwargs: dict[str, Any], expected_bytes: bytes):
+            """Test serialization of the object."""
+            obj = cls(**kwargs)
+            serialized_bytes = obj.serialize()
+            assert serialized_bytes == expected_bytes, f"{serialized_bytes} != {expected_bytes}"
+
+        @pytest.mark.parametrize(
+            ("kwargs", "expected_bytes"),
+            parameters,
+            ids=tuple(generate_name(kwargs, i) for i, (kwargs, _) in enumerate(parameters)),
+        )
+        def test_deserialization(self, kwargs: dict[str, Any], expected_bytes: bytes):
+            """Test deserialization of the object."""
+            buf = Buffer(expected_bytes)
+            obj = cls.deserialize(buf)
+            assert cls(**kwargs) == obj, f"{cls.__name__}({kwargs}) != {obj}"
+            assert buf.remaining == 0, f"Buffer has {buf.remaining} bytes remaining"
+
+        @pytest.mark.parametrize(
+            ("kwargs", "exc"),
+            validation_fail,
+            ids=tuple(generate_name(kwargs, i) for i, (kwargs, _) in enumerate(validation_fail)),
+        )
+        def test_validation(self, kwargs: dict[str, Any], exc: type[Exception]):
+            """Test validation of the object."""
+            with pytest.raises(exc):
+                cls(**kwargs)
```

Comment:
It'd be really nice if this also supported a custom pattern, so the generated tests could also check the exc msg.